torchrl/data/tensor_specs.py

```diff
@@ -3135,62 +3251,146 @@ def to_numpy(self, val: TensorDict, safe: bool = None) -> dict:
+            for spec, subvalue in zip(self._specs, value.unbind(self.dim)):
+                spec.type_check(subvalue)
+        else:
+            if isinstance(value, torch.Tensor) and isinstance(selected_keys, str):
+                value = {selected_keys: value}
+                selected_keys = [selected_keys]
+            for _key in self.keys():
+                if self[_key] is not None and _key in selected_keys:
+                    self[_key].type_check(value[_key], _key)
 
     def __repr__(self) -> str:
         sub_str = ",\n".join(
             [indent(f"{k}: {repr(item)}", 4 * " ") for k, item in self.items()]
         )
-        device_str = f"device={self._specs[0].device}"
-        shape_str = f"shape={self.shape}"
-        sub_str = ", ".join([sub_str, device_str, shape_str])
-        return (
-            f"LazyStackedCompositeSpec(\n{', '.join([sub_str, device_str, shape_str])})"
+        sub_str = indent(f"fields={{\n{', '.join([sub_str])}}}", 4 * " ")
+        exclusive_key_str = self.repr_exclusive_keys()
+        device_str = indent(f"device={self._specs[0].device}", 4 * " ")
+        shape_str = indent(f"shape={self.shape}", 4 * " ")
+        stack_dim = indent(f"stack_dim={self.dim}", 4 * " ")
+        string = ",\n".join(
+            [sub_str, exclusive_key_str, device_str, shape_str, stack_dim]
+        )
+        return f"LazyStackedCompositeSpec(\n{string})"
+
+    def repr_exclusive_keys(self):
+        keys = set(self.keys())
+        exclusive_keys = [
+            ",\n".join(
+                [
+                    indent(f"{k}: {repr(spec[k])}", 4 * " ")
+                    for k in spec.keys()
+                    if k not in keys
+                ]
+            )
+            for spec in self._specs
+        ]
+        exclusive_key_str = ",\n".join(
+            [
+                indent(f"{i} ->\n{line}", 4 * " ")
+                for i, line in enumerate(exclusive_keys)
+                if line != ""
+            ]
         )
 
+        return indent(f"exclusive_fields={{\n{exclusive_key_str}}}", 4 * " ")
+
+    def is_in(self, val) -> bool:
+        for spec, subval in zip(self._specs, val.unbind(self.dim)):
+            if not spec.is_in(subval):
+                return False
+        return True
+
     def encode(
         self, vals: Dict[str, Any], ignore_device: bool = False
     ) -> Dict[str, torch.Tensor]:
-        pass
+        raise NotImplementedError
 
-    def __delitem__(self, key):
-        pass
+    def __delitem__(self, key: NestedKey):
+        """Deletes a key only if present in all stacked specs."""
+        at_least_one_deletion = False
+        for spec in self._specs:
+            try:
+                del spec[key]
+                at_least_one_deletion = True
+            except KeyError:
+                continue
+        if not at_least_one_deletion:
+            raise KeyError(
+                f"Key {key} must be present in at least one of the stacked specs"
+            )
```

Comment:
this is problematic, because you raise an exception but do not repopulate the spec.
If I do

```
try:
    del spec[some_partial_key]
except:
    pass
```
I will end up with a partially depopulated spec

Reply:
wait but I only raise the exception if no deletion has occurred.
` if not at_least_one_deletion:` will mean nothing was touched

Reply:
Got it you're right, that should be ok
but the docstrings were misleading (from what I read it seemed to indicate that the key had to be everywhere)

Reply:
oh you are right, the docstirng if very wrong, thanks!